future/atomicTransactionComposer.go

```diff
@@ -0,0 +1,578 @@
+	 */
+	ReturnValue interface{}
+	/** If the SDK was unable to decode a return value, the error will be here. */
+	DecodeError error
+}
+
+type AddMethodCallParams struct {
+	/** The ID of the smart contract to call */
+	AppID uint64
+	/** The method to call on the smart contract */
+	Method Method
+	/** The arguments to include in the method call. If omitted, no arguments will be passed to the method. */
+	MethodArgs []interface{}
+	/** The address of the sender of this application call */
+	Sender types.Address
+	/** Transactions params to use for this application call */
+	SuggestedParams types.SuggestedParams
+	/** The OnComplete action to take for this application call. If omitted, OnApplicationComplete.NoOpOC will be used. */
+	OnComplete types.OnCompletion
+	/** The note value for this application call */
+	Note []byte
+	/** The lease value for this application call */
+	Lease [32]byte
+	/** If provided, the address that the sender will be rekeyed to at the conclusion of this application call */
+	RekeyTo types.Address
+}
+
+type AtomicTransactionComposerStatus = int
+
+const (
+	/** The atomic group is still under construction. */
+	BUILDING AtomicTransactionComposerStatus = iota
+
+	/** The atomic group has been finalized, but not yet signed. */
+	BUILT
+
+	/** The atomic group has been finalized and signed, but not yet submitted to the network. */
+	SIGNED
+
+	/** The atomic gorup has been finalized, signed, and submitted to the network. */
+	SUBMITTED
+
+	/** The atomic group has been finalized, signed, submitted, and successfully committed to a block. */
+	COMMITTED
+)
+
+type transactionContext struct {
+	// The main transaction.
+	txn types.Transaction
+
+	// The corresponding signer responsible for producing the signed transaction.
+	signer TransactionSigner
+
+	// The corresponding Method constructed from information passed into atc.AddMethodCall().
+	method *Method
+
+	// The raw signed transaction populated after invocation of atc.GatherSignatures().
+	stxBytes []byte
+}
+
+func (txContext *transactionContext) txID() string {
+	return crypto.GetTxID(txContext.txn)
+}
+
+func (txContext *transactionContext) isMethodCallTx() bool {
+	return txContext.method != nil
+}
+
+/** The maximum size of an atomic transaction group. */
+const MAX_GROUP_SIZE = 16
+
+/** A class used to construct and execute atomic transaction groups */
+type AtomicTransactionComposer struct {
+	/** The current status of the composer. The status increases monotonically. */
+	status AtomicTransactionComposerStatus
+
+	/** The transaction contexts in the group with their respective signers. If status is greater then
+	 *  BUILDING then this slice cannot change.
+	 */
+	txContexts []transactionContext
+
+	// caches that will be populated when respective contents are calculated
+	cachedTxWithSigners []TransactionWithSigner
+	cachedStxs          [][]byte
+	cachedTxIDs         []string
```

Comment:
I have some thoughts on these cache objects:
* `cachedTxWithSigners`: this can be calculated directly from `txContexts`, so I don't think caching this is worth the extra complexity
* `cachedStxs`: the signed transactions are already cached in the `stxBytes` field in `transactionContext`, so I don't see the point in also having this array
* `cachedTxIDs`: this is a benefitial cache since it avoids calculating more hashes than necessary, but I think it would be better implemented as a field in `transactionContext`. Then you could make `transactionContext.txID()` returned the cached version if it exists, otherwise calculate it. As currently implemented, the call on line 526 bypasses the cache, but it would be properly handled by this suggestion.

Reply:
The initial purpose of the caches was to avoid having to iterate over the txContexts and accumulate the extracted values. Maybe extraction/accumulation isn't expensive enough to justify this optimization?

Reply:
> Maybe extraction/accumulation isn't expensive enough to justify this optimization?

That's exactly my thought. Especially since transaction composers can have at most 16 transactions, the cost of extracting these fields is pretty minor.

Reply:
Makes sense, caches seem like overkill here.